### PR TITLE
Add safety rail for register_children method

### PR DESCRIPTION
**Pull request by PoignardAzur**
This gives a beginner-friendly error in cases where Masonry detects that `register_children()` was improperly implemented.

### CODE REVIEW

File: masonry/src/passes/update.rs [hunk 1]
```diff
@@ -492,6 +493,18 @@ fn update_new_widgets(
         // The widget will call `RegisterCtx::register_child` on all its children,
         // which will add the new widgets to the arena.
         widget.item.register_children(&mut ctx);
+
+        #[cfg(debug_assertions)]
+        for child_id in widget.item.children_ids() {
+            if widget.children.get_child(child_id.to_raw()).is_none() {
+                panic!(
+                    "Error in '{}' #{}: method register_children() did not call RegisterCtx::register_child() on child #{} returned by children_ids()",
```
Comment (waywardmonkeys) [inline masonry/src/passes/update.rs:501]: Is this string long enough that it might break `cargo fmt`?

**Review by waywardmonkeys — COMMENTED** [review]
When would someone want the implementation of register_children to do anything other than register each of the things in children_ids?

Curious why there are essentially two sources of truth here. This seems like a hidden invariant that isn't yet otherwise documented clearly.

Maybe an addition to the doc comment on `Widget::register_children` or `Widget::children_ids` would be good to have here.

**Comment by PoignardAzur** [comment]
> When would someone want the implementation of register_children to do anything other than register each of the things in children_ids?

Never, hence why we have a safety rail which checks that they're the same.

I'm still thinking about how to document this, but basically `children_ids` acts as the source of truth, and safety rails will give error messages that assume that the other method that doesn't match `children_ids` is wrong. (In my experience, `children_ids` methods are very easy to write and people never get them wrong. On the other hand, it's easy to forget to iterate over children in layout or register_children.)

> Curious why there are essentially two sources of truth here.

Basically the problem is that children_ids returns, well, ids, but when we want to add new widgets to the tree (the purpose of `register_children`), we need access to WidgetPods. In principle we could have a `children()` method that returns `Vec<WidgetPod<dyn Widget>>` or something, but that adds a lot of complications.

File: masonry/src/passes/update.rs (hunk 1 above)
Comment (PoignardAzur) [reply masonry/src/passes/update.rs:501]: It passes CI, so presumably not.

Comment (waywardmonkeys) [reply masonry/src/passes/update.rs:501]: Nah, `cargo fmt` silently fails on that.

Comment (PoignardAzur) [reply masonry/src/passes/update.rs:501]: What do you mean by silently fails? As in, fail to break the line?

Comment (waywardmonkeys) [reply masonry/src/passes/update.rs:501]: Usually it will also fail to format some of the surrounding code too!

You can detect this by adding this to `rustfmt.toml`, but they're not stable, so have to use `cargo +nightly fmt`:

```
error_on_line_overflow = true
error_on_unformatted = true
```

**Comment by waywardmonkeys** [comment]
If there's no rush, might as well see about the comment improvements before approving rather than have to remember that this needs further work again.

**Comment by PoignardAzur** [comment]
Oh, I thought you were talking about the formatting thing.

**Comment by PoignardAzur** [comment]
I've updated the documentation to make it a bit clearer.

**Review by waywardmonkeys — APPROVED** [review]
Much better with the doc comments.

You could link to the methods, but either way...

File: masonry/src/passes/update.rs [hunk 2]
```diff
@@ -498,10 +499,42 @@ fn update_new_widgets(
...
         };
         // The widget will call `RegisterCtx::register_child` on all its children,
         // which will add the new widgets to the arena.
         widget.item.register_children(&mut ctx);
+
+        #[cfg(debug_assertions)]
+        {
+            let children_ids = widget.item.children_ids();
+            for child_id in ctx.registered_ids {
+                if !children_ids.contains(&child_id) {
```
Comment (DJMcNab) [inline masonry/src/passes/update.rs:513]: I think that if we used `find` and `swap_remove` here, we would get the second check for free (just any items remaining in `children_ids`).

File: masonry/src/widget/widget.rs [hunk 3]
```diff
@@ -110,6 +111,9 @@ pub trait Widget: AsAny {
     /// the size of non-flex widgets first, to determine the amount of space
     /// available for the flex widgets.
     ///
+    /// Forgetting to visit children is a logic error and may lead to debug panics.
```
Comment (DJMcNab) [inline masonry/src/widget/widget.rs:114]: I don't really like the term "debug panics" - it's a bit oblique

File: masonry/src/passes/update.rs (hunk 2 above)
Comment (PoignardAzur) [reply masonry/src/passes/update.rs:513]: I'd rather have code be maximally clear.

What I'd like to do is add an automatic escape hatch for cases where `children_ids` is large, so we can skip checking those (under the assumption that if there's a bug in child handling, it probably shows up with a small number of children too).

**Comment by PoignardAzur** [comment]
"Panics in dev mode", then?

**Comment by DJMcNab** [comment]
Maybe "trigger debug assertions"
